prov/tcp/src/tcpx_progress.c

```diff
@@ -129,3 +154,39 @@ void tcpx_progress(struct util_ep *util_ep)
 	fastlock_release(&ep->queue_lock);
 	return;
 }
+
+static int tcpx_try_func(void *util_ep)
+{
+	/* nothing to do here. When endpoints
+	 have incoming data, cq drives progress*/
+	return FI_SUCCESS;
+}
+
+int tcpx_progress_ep_add(struct tcpx_ep *ep)
+{
+	if (ep->util_ep.rx_cq->wait) {
+		ofi_wait_fd_add(ep->util_ep.rx_cq->wait, ep->conn_fd,
+				tcpx_try_func, (void *)&ep->util_ep, NULL);
+
+		ofi_atomic_inc32(&ep->ref_cnt);
+	}
+ 	return FI_SUCCESS;
+}
+
+void tcpx_progress_ep_del(struct tcpx_ep *ep)
+{
+	fastlock_acquire(&ep->cm_state_lock);
+	if (ep->cm_state == TCPX_EP_CONNECTING) {
+		fastlock_release(&ep->cm_state_lock);
+		return;
+	}
+	fastlock_release(&ep->cm_state_lock);
```

Comment:
When the state is still connecting, we just bail out skipping ep delete. Once we are in some other state, we won't go back to that state again. So state going back to "TCPX_EP_CONNECTING" after releasing lock would not be an issue.

Reply:
It's not clear what the lock is protecting against is the main issue.  The state could change to CONNECTING immediately after releasing the lock, for example.

Reply:
I have added this check to prevent calling ep_del before ep_add ( i.e., before the connection was established).  

But since the ep_del gets called during ep close and conn management happens in another thread, there might be a race condition where ep_close might get the lock first and ep_add later. Thus ending up with closed socket in cq->wait object.